federationsender/queue/destinationqueue.go

```diff
@@ -33,92 +34,196 @@ import (
 // ensures that only one request is in flight to a given destination
 // at a time.
 type destinationQueue struct {
-	rsProducer  *producers.RoomserverProducer
-	client      *gomatrixserverlib.FederationClient
-	origin      gomatrixserverlib.ServerName
-	destination gomatrixserverlib.ServerName
-	running     atomic.Bool
-	// The running mutex protects sentCounter, lastTransactionIDs and
-	// pendingEvents, pendingEDUs.
-	runningMutex       sync.Mutex
-	sentCounter        int
-	lastTransactionIDs []gomatrixserverlib.TransactionID
-	pendingEvents      []*gomatrixserverlib.HeaderedEvent
-	pendingEDUs        []*gomatrixserverlib.EDU
-	pendingInvites     []*gomatrixserverlib.InviteV2Request
+	rsProducer         *producers.RoomserverProducer        // roomserver producer
+	client             *gomatrixserverlib.FederationClient  // federation client
+	origin             gomatrixserverlib.ServerName         // origin of requests
+	destination        gomatrixserverlib.ServerName         // destination of requests
+	running            atomic.Bool                          // is the queue worker running?
+	wakeup             chan bool                            // wakes up a sleeping worker
+	statistics         *types.ServerStatistics              // statistics about this remote server
+	idleCounter        atomic.Uint32                        // how many ticks have we done nothing?
+	runningMutex       sync.RWMutex                         // protects the below
+	lastTransactionIDs []gomatrixserverlib.TransactionID    // protected by runningMutex
+	pendingPDUs        []*gomatrixserverlib.HeaderedEvent   // protected by runningMutex
+	pendingEDUs        []*gomatrixserverlib.EDU             // protected by runningMutex
+	pendingInvites     []*gomatrixserverlib.InviteV2Request // protected by runningMutex
+}
+
+// Start the destination queue if it needs to be started, or
+// otherwise signal to it that it should wake up from sleep.
+func (oq *destinationQueue) wake() {
+	if !oq.running.Load() {
+		go oq.backgroundSend()
+	} else {
+		select {
+		case oq.wakeup <- true:
```

Comment:
Why is this in a `select`?

Reply:
We don't actually know that the worker is sleeping at that point. I think we just want a best-effort "wake up" if it just happens to be rather than waiting for it.